File: sdk/python/examples/langgraph/24_map_reduce.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Map-Reduce — generate documents, summarize each, then reduce to a final report.

Demonstrates:
    - Generate multiple documents about a topic as a tool
    - Summarize each document individually as a tool
    - Reduce all summaries into a cohesive final report as a tool
    - The LLM orchestrates the map-reduce pipeline server-side

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def generate_documents(topic: str) -> str:
    """Generate 3 short document snippets about the topic.

    Returns the snippets as a numbered list.

    Args:
        topic: The topic to generate document snippets about.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", (
            "Generate 3 short text snippets (each 2-3 sentences) about the given topic. "
            "Format as a numbered list:\n1. ...\n2. ...\n3. ..."
        )),
        ("human", "Topic: {topic}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"topic": topic})
    return response.content.strip()


@tool
def summarize_document(topic: str, document: str) -> str:
    """Summarize a single document snippet in one concise sentence.

    Args:
        topic: The overall topic for context.
        document: The document snippet to summarize.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Summarize this text in one concise sentence."),
        ("human", "Topic: {topic}\n\nText: {document}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"topic": topic, "document": document})
    return response.content.strip()


@tool
def reduce_summaries(topic: str, summaries: str) -> str:
    """Combine multiple document summaries into a cohesive 2-3 sentence final report.

    Args:
        topic: The topic being reported on.
        summaries: All document summaries combined (bullet points or numbered list).
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", (
            "You are a report writer. Given the topic and a list of summaries, "
            "write a cohesive 2-3 sentence final report."
        )),
        ("human", "Topic: {topic}\n\nSummaries:\n{summaries}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"topic": topic, "summaries": summaries})
    return response.content.strip()


MAP_REDUCE_SYSTEM = """You are a map-reduce research agent.

For each research topic:
1. Call generate_documents to create 3 document snippets
2. Call summarize_document once for EACH snippet (map phase)
3. Collect all summaries and call reduce_summaries to produce the final report (reduce phase)
4. Return the final report to the user

Make sure to summarize each document snippet individually before reducing.
"""

graph = create_agent(
    llm,
    tools=[generate_documents, summarize_document, reduce_summaries],
    name="map_reduce_agent",
    system_prompt=MAP_REDUCE_SYSTEM,
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(graph, "renewable energy breakthroughs in 2024")
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/24-map-reduce.ts
/**
 * Map-Reduce -- fan-out to parallel workers then aggregate results.
 *
 * Demonstrates:
 *   - Fan-out pattern for parallel processing via Send API
 *   - Processing multiple items concurrently
 *   - Reducing parallel results into a single final answer
 *   - Practical use case: analyzing multiple documents simultaneously
 *
 * In production you would use:
 *   import { StateGraph, START, END } from '@langchain/langgraph';
 *   import { Send } from '@langchain/langgraph';
 */

import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// State types
// ---------------------------------------------------------------------------
interface OverallState {
  topic: string;
  documents: string[];
  summaries: string[];
  finalReport: string;
}

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function generateDocuments(state: OverallState): Partial<OverallState> {
  // Mock: generate 3 document snippets about the topic
  const docs = [
    `Solar panel efficiency has surpassed 30% in lab conditions, with perovskite cells leading the way. Commercial deployment costs have dropped 89% since 2010.`,
    `Wind energy capacity additions reached 117 GW globally in 2024. Offshore wind is seeing rapid growth, particularly in Northern Europe and Asia Pacific.`,
    `Battery storage technology breakthroughs include solid-state batteries with 500 Wh/kg energy density. Grid-scale storage installations doubled year-over-year.`,
  ];
  return { documents: docs };
}

function summarizeDoc(doc: string, topic: string): string {
  // Mock: summarize a single document
  const words = doc.split(' ').slice(0, 8).join(' ');
  return `Summary of "${words}...": Key developments in ${topic} sector.`;
}

function fanOutAndSummarize(state: OverallState): Partial<OverallState> {
  // Process each document (in production, these would be parallel via Send)
  const summaries = state.documents.map((doc) =>
    summarizeDoc(doc, state.topic),
  );
  return { summaries };
}

function reduceSummaries(state: OverallState): Partial<OverallState> {
  const bulletPoints = state.summaries.map((s) => `  - ${s}`).join('\n');
  const finalReport =
    `Final Report on "${state.topic}":\n` +
    `Analyzed ${state.documents.length} documents.\n\n` +
    `Key Findings:\n${bulletPoints}\n\n` +
    `Conclusion: Significant progress across all renewable energy sectors with ` +
    `cost reductions and efficiency improvements driving rapid adoption.`;
  return { finalReport };
}

// ---------------------------------------------------------------------------
// Mock compiled graph
// ---------------------------------------------------------------------------
const graph = {
  name: 'map_reduce_agent',

  invoke: async (input: Record<string, unknown>) => {
    const topic = (input.input as string) ?? '';
    let state: OverallState = {
      topic,
      documents: [],
      summaries: [],
      finalReport: '',
    };

    state = { ...state, ...generateDocuments(state) };
    state = { ...state, ...fanOutAndSummarize(state) };
    state = { ...state, ...reduceSummaries(state) };

    return { output: state.finalReport };
  },

  getGraph: () => ({
    nodes: new Map([
      ['__start__', {}],
      ['generate_documents', {}],
      ['summarize_doc', {}],
      ['reduce', {}],
      ['__end__', {}],
    ]),
    edges: [
      ['__start__', 'generate_documents'],
      ['generate_documents', 'summarize_doc'],
      ['summarize_doc', 'reduce'],
      ['reduce', '__end__'],
    ],
  }),

  nodes: new Map([
    ['generate_documents', {}],
    ['summarize_doc', {}],
    ['reduce', {}],
  ]),

  stream: async function* (input: Record<string, unknown>) {
    const topic = (input.input as string) ?? '';
    let state: OverallState = {
      topic,
      documents: [],
      summaries: [],
      finalReport: '',
    };

    state = { ...state, ...generateDocuments(state) };
    yield ['updates', { generate_documents: { documents: state.documents } }];

    state = { ...state, ...fanOutAndSummarize(state) };
    yield ['updates', { summarize_doc: { summaries: state.summaries } }];

    state = { ...state, ...reduceSummaries(state) };
    yield ['updates', { reduce: { finalReport: state.finalReport } }];

    yield ['values', { output: state.finalReport }];
  },
};

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(
      graph,
      'renewable energy breakthroughs in 2024',
    );
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
